server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationBalancingRoundSummaryService.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Manages the lifecycle of a series of {@link BalancingRoundSummary} results from allocation balancing rounds and creates reports thereof.
+ * Reporting balancer round summary results will provide information with which to do cost-benefit analyses of the work that shard
+ * allocation rebalancing executes.
+ *
+ * Any successfully added summary via {@link #addBalancerRoundSummary(BalancingRoundSummary)} will eventually be collected/drained and
+ * reported. This should still be done in the event of the node stepping down from master, on the assumption that all summaries are only
+ * added while master and should be drained for reporting. There is no need to start/stop this service with master election/stepdown because
+ * balancer rounds will no longer be supplied when not master. It will simply drain the last summaries and then have nothing more to do.
+ * This does have the tradeoff that non-master nodes will run a task to check for summaries to report every
+ * {@link #BALANCER_ROUND_SUMMARIES_LOG_INTERVAL_SETTING} seconds.
+ */
+public class AllocationBalancingRoundSummaryService {
+
+    /** Turns on or off balancing round summary reporting. */
+    public static final Setting<Boolean> ENABLE_BALANCER_ROUND_SUMMARIES_SETTING = Setting.boolSetting(
+        "cluster.routing.allocation.desired_balance.enable_balancer_round_summaries",
+        false,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    /** Controls how frequently in time balancer round summaries are logged. If less than zero, effectively disables reporting. */
+    public static final Setting<TimeValue> BALANCER_ROUND_SUMMARIES_LOG_INTERVAL_SETTING = Setting.timeSetting(
+        "cluster.routing.allocation.desired_balance.balanace_round_summaries_interval",
+        TimeValue.timeValueSeconds(10),
+        TimeValue.ZERO,
```

Comment:
It says above
> If less than zero, effectively disables reporting.
but the minimum is zero (I think?)

Either comment or code is out of date I think